### PR TITLE
feat: utils for RBMap

### PR DESCRIPTION
**Pull request by EdAyers**
Some utils that I needed when writing code actions

- RBMap.mergeBy
- RBNode.map is useful for working with RBNodes in Json. Writing a proof that it is wf is more tricky.
- RBNode.toArray
- some doccstrings

### CODE REVIEW

File: src/Bootstrap/Data/RBMap.lean [hunk 1]
```diff
@@ -225,6 +225,28 @@ inductive WellFormed (cmp : α → α → Ordering) : RBNode α β → Prop wher
...
+@[specialize] def map {α : Type u} {β γ : α → Type v}
+  (f : {a : α} → β a → γ a)
+  : RBNode α β → RBNode α γ
+  | leaf => leaf
+  | node color lchild key val rchild => node color (lchild.map f) key (f val) (rchild.map f)
+
+end Map
+
+def toArray (n : RBNode α β) : Array (Sigma β) :=
+  n.fold (fun acc k v => acc.push ⟨k,v⟩) ∅
```
Comment (gebner) [inline src/Bootstrap/Data/RBMap.lean:246]:
```suggestion
  n.fold (init := ∅) fun acc k v => acc.push ⟨k,v⟩
```
The init argument is really in the wrong position.

File: src/Bootstrap/Data/RBMap.lean [hunk 2]
```diff
@@ -343,12 +371,25 @@ def maxDepth (t : RBMap α β cmp) : Nat :=
...
 
+/-- Merges the maps `t₁` and `t₂`, if a key `a : α` exists in both,
+then use `mergeFn a b₁ b₂` to produce the new merged value. -/
+def mergeBy (mergeFn : α → β → β → β) (t₁ t₂ : RBMap α β cmp) : RBMap α β cmp :=
+  t₂.fold (fun t₁ a b₂ =>
+    t₁.insert a <|
+      match t₁.find? a with
+      | some b₁ => mergeFn a b₁ b₂
+      | none => b₂
+  ) t₁
```
Comment (gebner) [inline src/Bootstrap/Data/RBMap.lean:388]:
```suggestion
  t₂.fold (init := t₁) fun t₁ a b₂ =>
    t₁.insert a <|
      match t₁.find? a with
      | some b₁ => mergeFn a b₁ b₂
      | none => b₂
```

Same as with the other issue I pointed out earlier.

**Comment by leodemoura** [comment]
@EdAyers The file has moved. Could you please fix the PR?

File: src/Lean/Data/RBMap.lean [hunk 3]
```diff
@@ -225,6 +225,28 @@ inductive WellFormed (cmp : α → α → Ordering) : RBNode α β → Prop wher
   | insertWff {n n' : RBNode α β} {k : α} {v : β k} : WellFormed cmp n → n' = insert cmp n k v → WellFormed cmp n'
   | eraseWff {n n' : RBNode α β} {k : α} : WellFormed cmp n → n' = erase cmp k n → WellFormed cmp n'
 
+section Map
+
+@[specialize] def mapM {α : Type v} {β γ : α → Type v} {M : Type v → Type v} [Applicative M]
+  (f : (a : α) → β a → M (γ a))
```
Comment (leodemoura) [inline src/Lean/Data/RBMap.lean:231]: Why `a` is explicit here, and implicit at `map`? I expected both to be implicit.
Is this a workaround for some kind of Lean bug/limitation?

Comment (EdAyers) [reply src/Lean/Data/RBMap.lean:231]: this was a mistake by me

**Pull request merged**